fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexPrefetchTest.java

```diff
@@ -0,0 +1,483 @@
+        RecordQueryPlan plan = RecordQueryComparatorPlan.from(List.of(planWithScan, planWithPrefetch), primaryKey(), 0, true);
+
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context, simpleMetadataHook);
+            try (RecordCursor<FDBQueriedRecord<Message>> cursor = recordStore.executeQuery(plan, null, executeProperties)) {
+                // Will throw exception if plans do not match
+                // This only compares the primary key and not all values
+                assertNotNull(cursor.asList().get());
+            }
+        }
+    }
+
+    @Test
+    void testIndexPrefetchWithComparatorPlanFails() throws Exception {
+        RecordQueryPlan planWithScan = plan(NUM_VALUES_LARGER_THAN_990, RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE);
+        RecordQueryPlan planWithPrefetch = plan(STR_VALUE_EVEN, RecordQueryPlannerConfiguration.IndexPrefetchUse.USE_INDEX_PREFETCH);
+        ExecuteProperties executeProperties = ExecuteProperties.SERIAL_EXECUTE;
+        RecordQueryPlan plan = RecordQueryComparatorPlan.from(List.of(planWithScan, planWithPrefetch), primaryKey(), 0, true);
+
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context, simpleMetadataHook);
+            try (RecordCursor<FDBQueriedRecord<Message>> cursor = recordStore.executeQuery(plan, null, executeProperties)) {
+                assertThrows(ExecutionException.class, () -> cursor.asList().get());
+            }
+        }
+    }
+
+    @Test
+    void testIndexPrefetchReadYourWriteFails() throws Exception {
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context, simpleMetadataHook);
+            // Save record (don't commit)
+            TestRecords1Proto.MySimpleRecord.Builder recBuilder = TestRecords1Proto.MySimpleRecord.newBuilder();
+            recBuilder.setRecNo(1000);
+            recBuilder.setStrValueIndexed("even");
+            recBuilder.setNumValueUnique(1000);
+            recBuilder.setNumValue2(3);
+            recBuilder.setNumValue3Indexed(5);
+            recordStore.saveRecord(recBuilder.build());
+
+            // Execute the query (will fail because a record in memory cannot be processed by index prefetch)
+            RecordQueryPlan plan = plan(STR_VALUE_EVEN, RecordQueryPlannerConfiguration.IndexPrefetchUse.USE_INDEX_PREFETCH);
+            ExecuteProperties executeProperties = ExecuteProperties.SERIAL_EXECUTE;
+            RecordCursor<FDBQueriedRecord<Message>> cursor = recordStore.executeQuery(plan, null, executeProperties);
+
+            // This should fail when index prefetch is fully supported. Since for now the API_VERSION is 630, prefetch mode
+            // will fall back to index scan, which would not fail in this case.
+            assertNotNull(cursor.getNext());
+            // assertThrows(RecordCoreException.class, () -> cursor.getNext());
```

Comment:
In theory, this assert could be guarded by API version if you want to make assertions about behavior at different API versions. Something like:

```java
if (context.isAPIVersionAtLeast(APIVersion.API_VERSION_7_1)) {
    assertThrows(RecordCoreException.class, () -> cursor.getNext());
} else {
    assertNotNull(cursor.getNext());
}
```

There are a few other test asserts that maybe could be written this way to prepare this test fixture for running at multiple API versions

Reply:
Nice! Done.